Maintain sort order after collapsing a context

finkef opened the pull request:
Fixes #2091.

This PR re-applies the sort preference (if any exists) on collapseContext after it has moved to the parent context.

raineorshine reviewed (changes requested):
We should be able to take a more efficient approach here [like in 2160](https://github.com/cybersemics/em/pull/2160).

In this case, I would suggest sorting the parent context if it is unsorted and then outdenting the children as usual.

finkef commented:
Rebased on main and changed to preferring `getSortedRank` for inserting into the sorted parent context.

Curious if this is technically more efficient since we now have 1 `sort` over a smaller space + n `getSortedRank`, instead of one `sort` over the combined space. In reality, there will most likely be no measurable difference.

raineorshine commented on src/actions/collapseContext.ts:
```diff
@@ -69,13 +76,16 @@ const collapseContext = (state: State, { at }: Options) => {
       newValue: createId(), // unique value
       path: simplePath,
     }),
+    // Sort parent context if sort preference exists
+    sortId ? sort(parentId) : null,
```
Only sort the parent if it's not already sorted (i.e. check the sort status of the parent with [getSortPreference](https://github.com/cybersemics/em/blob/a27f56c62a1a16d0b31435201cd77aed7ccd156e/src/selectors/getSortPreference.ts)).

raineorshine reviewed (changes requested):
This approach does read better, thanks.

This is working for the original case, but it's broken for a different case though. Can you investigate? I think it's interacting with a feature that is meant to switch to manual sort when the user drags and drops a thought into a different order.

## Steps to Reproduce

```ts
- d
- c
- b
  - =sort
    - Alphabetical
      - Asc
  - a
  - e
  - g
- f
```

Set the cursor on `b` and activate `collapseContext`.

## Current Behavior

5eee62c77d

```
- d
- a
- c
- e
- g
- f
```

## Expected Behavior

```
- =sort
  - Alphabetical
    - Asc
- a
- c
- d
- e
- f
- g
```

finkef commented:
> This approach does read better, thanks. And even if it is asymptotically slower, it updates fewer thoughts.
> 
> This is working for the original case, but it's broken for a different case unfortunately. Can you investigate? I think it's interacting with a feature that is meant to switch to manual sort when the user drags and drops a thought into a different order.

Fixed by first moving up the sort preference and sorting (if necessary), then other thoughts. Also added a test case to cover this scenario.

raineorshine commented:
Looks good, and thanks for the added test.
